indico/modules/categories/controllers/management.py

```diff
@@ -102,15 +103,34 @@ def _process(self):
 
 class RHManageCategorySettings(RHManageCategoryBase):
     def _process(self):
-        defaults = FormDefaults(self.category,
-                                meeting_theme=self.category.default_event_themes['meeting'],
-                                lecture_theme=self.category.default_event_themes['lecture'])
+        additional_keys = ('google_wallet_application_credentials', 'google_wallet_issuer_name',
+                           'google_wallet_issuer_id')
+        kwargs = {
+            'meeting_theme': self.category.default_event_themes['meeting'],
+            'lecture_theme': self.category.default_event_themes['lecture']}
+
+        if config.ENABLE_GOOGLE_WALLET:
+            kwargs['google_wallet_enabled'] = self.category.google_wallet_settings
+            if self.category.google_wallet_settings:
+                kwargs.update({key: self.category.google_wallet_settings[key] for key in additional_keys})
+        defaults = FormDefaults(self.category, **kwargs)
         form = CategorySettingsForm(obj=defaults, category=self.category)
         icon_form = CategoryIconForm(obj=self.category)
         logo_form = CategoryLogoForm(obj=self.category)
 
         if form.validate_on_submit():
-            update_category(self.category, form.data, skip={'meeting_theme', 'lecture_theme'})
+            new_dict = form.data
+            new_dict.setdefault('google_wallet_settings', {})
+            if config.ENABLE_GOOGLE_WALLET:
+                if new_dict.pop('google_wallet_enabled'):
+                    for key in additional_keys:
+                        new_dict['google_wallet_settings'][key] = new_dict.pop(key, {})
+                else:
+                    for key in additional_keys:
+                        del new_dict[key]
```

Comment:
Check the (custom) `@generated_data` decorator. I think you can do something like this in the `CategorySettingsForm` (UNTESTED):

```python
@generated_data
def google_wallet_settings(self):
    google_wallet_keys = (
        'google_wallet_application_credentials',
        'google_wallet_issuer_name',
        'google_wallet_issuer_id',
    )
    if 'google_wallet_enabled' in self and self.google_wallet_enabled.data:
        return {k: v for k, v in self.data.items() if k in google_wallet_keys}
    else:
        # might need to expose `category` via the form object, since i don't think you can
        # have a generated_data property return nothing (or rather doing so would overwrite
        # the existing data)...
        return self.category.google_wallet_settings
```

Reply:
Ffor the category I passed: form.category = self.category
I tried but I cannot make it work:

`ValueError: Category has no attribute 'google_wallet_enabled'`

Reply:
It's cleaner to add this to the form and pass the category as a kwarg then:

```py
    def __init__(self, *args, category, **kwargs):
        super().__init__(*args, **kwargs)
        self.category = category

```

About the error: A traceback would be good. I do not see anything accessing `self.category.google_wallet_enabled` in my code...

Reply:
ValueError
ValueError: Category has no attribute 'google_wallet_enabled'

Traceback (most recent call last)
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/sentry_sdk/integrations/flask.py", line 89, in sentry_patched_wsgi_app
return SentryWsgiMiddleware(lambda *a, **kw: old_app(self, *a, **kw))(
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/sentry_sdk/integrations/wsgi.py", line 115, in __call__
reraise(*_capture_exception(hub))
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/sentry_sdk/_compat.py", line 115, in reraise
raise value
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/sentry_sdk/integrations/wsgi.py", line 108, in __call__
rv = self.app(
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/sentry_sdk/integrations/flask.py", line 89, in <lambda>
return SentryWsgiMiddleware(lambda *a, **kw: old_app(self, *a, **kw))(
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/app.py", line 1478, in __call__
return self.wsgi_app(environ, start_response)
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/app.py", line 1458, in wsgi_app
response = self.handle_exception(e)
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/app.py", line 1455, in wsgi_app
response = self.full_dispatch_request()
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/app.py", line 869, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/app.py", line 759, in handle_user_exception
return self.ensure_sync(handler)(e)
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/app.py", line 867, in full_dispatch_request
rv = self.dispatch_request()
File "/Users/giorgiopieretti/Work/indico-vanilla/vanilla-env/lib/python3.9/site-packages/flask/app.py", line 852, in dispatch_request
return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/flask/util.py", line 80, in wrapper
return obj().process()
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/rh.py", line 292, in process
res = self._do_process()
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/web/rh.py", line 263, in _do_process
rv = self._process()
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/modules/categories/controllers/management.py", line 133, in _process
update_category(self.category, new_dict, skip={'meeting_theme', 'lecture_theme'},
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/util/signals.py", line 111, in _wrapper
return func(*args, **kwargs)
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/modules/categories/operations.py", line 55, in update_category
changes = category.populate_from_dict(data, skip=skip)
File "/Users/giorgiopieretti/Work/indico-vanilla/src/indico/core/db/sqlalchemy/util/models.py", line 213, in populate_from_dict
raise ValueError(f"{cls.__name__} has no attribute '{key}'")
ValueError: Category has no attribute 'google_wallet_enabled'

Reply:
ah, because it still tries to set this data.

```
update_category(self.category, new_dict, skip={'meeting_theme', 'lecture_theme'},
```

Simply add those fields to the `skip` list there, or override the `data` property in the form and filter those fields out from that dict.

smart option (if it works, untested):

```py
    @property
    def data(self):
        generated_data = self.generated_data
        return {k: v for k, v in super().data.items() if k not in generated_data}
```

Reply:
Sorry, I do not get it... where should I add the "smart option" code?
on the other hand... I'm testing with the rest of added code still there... is it ok?

```
class RHManageCategorySettings(RHManageCategoryBase):
    def _process(self):
        additional_keys = ('google_wallet_application_credentials', 'google_wallet_issuer_name',
                           'google_wallet_issuer_id')
        kwargs = {
            'meeting_theme': self.category.default_event_themes['meeting'],
            'lecture_theme': self.category.default_event_themes['lecture']}

        if config.ENABLE_GOOGLE_WALLET:
            kwargs['google_wallet_enabled'] = self.category.google_wallet_settings
            if self.category.google_wallet_settings:
                kwargs.update({key: self.category.google_wallet_settings[key] for key in additional_keys})
        defaults = FormDefaults(self.category, **kwargs)
        form = CategorySettingsForm(obj=defaults, category=self.category)
        icon_form = CategoryIconForm(obj=self.category)
        logo_form = CategoryLogoForm(obj=self.category)

        if form.validate_on_submit():
            form.category = self.category
            new_dict = form.data
            update_category(self.category, new_dict, skip={'meeting_theme', 'lecture_theme'},
                            _extra_log_fields={'google_wallet_settings': 'Google Wallet configuration settings'})
            self.category.default_event_themes = {
                'meeting': form.meeting_theme.data,
                'lecture': form.lecture_theme.data
            }
            flash(_('Category settings saved!'), 'success')
            return redirect(url_for('.manage_settings', self.category))
        else:
            if self.category.icon_metadata:
                icon_form.icon.data = self.category
            if self.category.logo_metadata:
                logo_form.logo.data = self.category
        return WPCategoryManagement.render_template('management/settings.html', self.category, 'settings', form=form,
                                                    icon_form=icon_form, logo_form=logo_form)
```

Reply:
that code would go into the CategorySettingsForm

Reply:
OK, tnx! :)
I added the code, no more errors BUT if I change the Google Wallet data and save, my changes are not saved...

Reply:
I'll be on leave from tomorrow 'til the end of the week... I propose to keep the code as it is (that is working fine) or pls directly apply and test the changes you proposed, thank you! :)

Reply:
i'm testing it now, so i might be able to get in some feedback before you go on holidays. otherwise i can look into fixing (minor) things myself to hopefully get this merged soon :)

Reply:
Done: Optimisation on RegForm